### PR TITLE
[Coupons in Orders]Provide a better handling for adding coupons to orders errors.

### LINKED ISSUE #10041: [Orders with Coupons]Handle coupon cannot be added to order

### PR DESCRIPTION
**Pull request by toupper**
Closes: #10041

## Description

With this PR we provide a better handling for errors when adding coupons to orders. We do this through two tasks:

1. Disable the Add Coupon row in the Order Details screen if there are no products, as it will surely fail.
2. Show a proper message to the user and reset the Order coupon if adding the coupon fails because it cannot be added to the order.

### Known problem

Currently, if we don't handle the error when we sync an order, we show `Unable to load taxes for order`, which might not be the real problem behind the error. This however shouldn't affect us here as we handle the error.

## Testing instructions

1. Go to orders and tap to create a new one
2. Add a product
3. Try to add a coupon not applicable to the added product. It should fail with a specific error. The coupon should be removed from the order.
4. Add a coupon applicable to the product. It should sync succesfully.

## Screenshots

https://github.com/woocommerce/woocommerce-ios/assets/1864060/2c3b5a3d-9754-4487-b50f-0318199bfa2d

---
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/EditableOrderViewModelTests.swift [hunk 1]
```diff
@@ -2,6 +2,7 @@ import XCTest
 @testable import WooCommerce
 import Yosemite
 import WooFoundation
+import Networking
```
Comment (iamgabrielma) [inline WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/EditableOrderViewModelTests.swift:5]: Should we use a different approach to test this (like stubbing the DotComError perhaps?) instead of importing Networking into WooCommerce ?

```suggestion
```

``` diff
enum DotcomErrorStub: Error {
  case unknown(code: String, message: String?)
}
        
let dotComError = DotcomErrorStub.unknown(code: "woocommerce_rest_invalid_coupon", message: "")

... 
onCompletion(.failure(dotComError))
expectation.fulfill()

```

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/OrderPaymentSection.swift [hunk 2]
```diff
@@ -105,13 +105,15 @@ struct OrderPaymentSection: View {
             TitleAndValueRow(title: viewModel.couponSummary ?? Localization.coupon, value: .content(viewModel.discountTotal), selectionStyle: .highlight) {
                 shouldShowCouponLineDetails = true
             }
+            .disabled(viewModel.shouldDisableAddingCoupons)
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/OrderPaymentSection.swift:108]: 💯 

Not part of this PR, but we could possibly consider apply the same modifier to "+ Add Shipping" and "+ Add Fees".

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift [hunk 3]
```diff
@@ -1425,6 +1436,14 @@ extension EditableOrderViewModel {
                 return false
             }
         }
+
+        private static func isCouponsError(_ error: Error) -> Bool {
+            if case .unknown(code: "woocommerce_rest_invalid_coupon", _) = error as? DotcomError {
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift:1442]: I see this was used already on a different method, but - related to the comment in the tests - , we seem to accept any sort of Error in this method, but then internally we're requiring a specific DotComError type, which requires us to import `enum Networking.DotcomError` and leak networking details into the business layer. Perhaps we should abstract this?

**Review by iamgabrielma — APPROVED** [review]
Tests well! My only concern is to access Networking both in the ViewModel and in the tests, since ideally this layer should not know about Networking.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift (hunk 3 above)
Comment (toupper) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift:1442]: That's an excellent point, thanks for commenting. 

If we want to avoid exposing the `DotcomError` in the WooCommerce target, we would have to parse and convert the `DotcomError` in `Yosemite` to some other error enum public which can be read from WooCommerce. 

The tradeoff of accessing the `DotcomError` in WooCommerce is that we don't have to do that extra task of parsing and duplicating the error, we can just pass along the retrieved one from `Networking`, which is not optimal but more straightforward. This is the general strategy in the app; we read the `DotcomError` in quite a few places through the WooCommerce target.

File: WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/EditableOrderViewModelTests.swift (hunk 1 above)
Comment (toupper) [reply WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/EditableOrderViewModelTests.swift:5]: Since we are actually importing that `Networking` `DotComError`in `EditableOrderViewModel`, I think its fine if we replicate that behavior and import `Networking` in `WooCommerce`. I changed the import a little bit to `import enum Networking.DotcomError` to make it more specific as in the `EditableOrderViewModel`.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift (hunk 3 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift:1442]: Gotcha! Indeed there seem to be quite a bit of usage of this across the app. I've opened an [issue here](https://github.com/woocommerce/woocommerce-ios/issues/10070) to keep track of it 👍
